File: structural-design-pattern/facade-pattern/facade.go
// Facade Pattern
// ==============
// Category: Structural Design Pattern
//
// Intent:
//   Provide a unified, simplified interface to a complex subsystem.
//   The facade hides internal complexity and coordinates multiple services.
//
// When to use:
//   - When a subsystem has many interacting types
//   - When clients only need simple operations from a complex system
//   - When you want to decouple clients from subsystem internals
//
// Key Participants:
//   - Facade: Simple API that delegates to subsystem types
//   - Subsystem types: Do the actual work
//   - Client: Uses only the facade

package main

import (
	"fmt"
	"time"
)

// ---------------------------------------------------------------------------
// Subsystem: Inventory
// ---------------------------------------------------------------------------

type InventoryService struct {
	stock map[string]int
}

func NewInventoryService() *InventoryService {
	return &InventoryService{
		stock: map[string]int{
			"LAPTOP-001":   50,
			"MOUSE-002":    200,
			"KEYBOARD-003": 75,
		},
	}
}

func (s *InventoryService) Reserve(productID string, quantity int) bool {
	available := s.stock[productID]
	if available >= quantity {
		s.stock[productID] -= quantity
		fmt.Printf("  [Inventory] Reserved %dx %s\n", quantity, productID)
		return true
	}
	return false
}

// ---------------------------------------------------------------------------
// Subsystem: Payment
// ---------------------------------------------------------------------------

type PaymentService struct{}

func (p *PaymentService) Charge(customerID string, amount float64) (bool, string) {
	fmt.Printf("  [Payment] Charged $%.2f to customer %s\n", amount, customerID)
	return true, fmt.Sprintf("PAY-%d", time.Now().UnixNano())
}

// ---------------------------------------------------------------------------
// Subsystem: Shipping
// ---------------------------------------------------------------------------

type ShippingService struct{}

func (s *ShippingService) CalculateCost(weightKg float64) float64 {
	return 5.99 + weightKg*1.50
}

func (s *ShippingService) CreateShipment(orderID, destination string) string {
	tracking := fmt.Sprintf("SHIP-%d", time.Now().UnixNano())
	fmt.Printf("  [Shipping] Created shipment for order %s to %s\n", orderID, destination)
	return tracking
}

// ---------------------------------------------------------------------------
// Subsystem: Notification
// ---------------------------------------------------------------------------

type NotificationService struct{}

func (n *NotificationService) SendEmail(email, subject string) {
	fmt.Printf("  [Notification] Email to %s: %s\n", email, subject)
}

// ---------------------------------------------------------------------------
// Facade
// ---------------------------------------------------------------------------

type OrderResult struct {
	Success  bool
	OrderID  string
	Total    float64
	Tracking string
	Error    string
}

type OrderFacade struct {
	inventory    *InventoryService
	payment      *PaymentService
	shipping     *ShippingService
	notification *NotificationService
}

func NewOrderFacade() *OrderFacade {
	return &OrderFacade{
		inventory:    NewInventoryService(),
		payment:      &PaymentService{},
		shipping:     &ShippingService{},
		notification: &NotificationService{},
	}
}

// PlaceOrder — one simple call replaces 5+ subsystem calls.
func (f *OrderFacade) PlaceOrder(customerID, email, productID string, quantity int, unitPrice float64, destination string) OrderResult {
	fmt.Printf("Processing order for %s...\n", customerID)

	if !f.inventory.Reserve(productID, quantity) {
		return OrderResult{Success: false, Error: "Product out of stock"}
	}

	shippingCost := f.shipping.CalculateCost(float64(quantity) * 0.5)
	total := unitPrice*float64(quantity) + shippingCost

	paid, _ := f.payment.Charge(customerID, total)
	if !paid {
		return OrderResult{Success: false, Error: "Payment failed"}
	}

	orderID := fmt.Sprintf("ORD-%d", time.Now().UnixNano())
	tracking := f.shipping.CreateShipment(orderID, destination)

	f.notification.SendEmail(email, "Order Confirmed!")

	return OrderResult{
		Success:  true,
		OrderID:  orderID,
		Total:    total,
		Tracking: tracking,
	}
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
func main() {
	facade := NewOrderFacade()

	result := facade.PlaceOrder("CUST-42", "alice@example.com", "LAPTOP-001", 2, 999.99, "New York, NY")
	fmt.Printf("\nOrder result: %+v\n", result)

	fmt.Println("\n--- Out of stock scenario ---")
	result2 := facade.PlaceOrder("CUST-99", "bob@example.com", "NONEXISTENT", 1, 49.99, "LA, CA")
	fmt.Printf("Order result: %+v\n", result2)
}

File: structural-design-pattern/facade-pattern/facade.rs
//! Facade Pattern
//! ==============
//! Category: Structural Design Pattern
//!
//! Intent:
//!   Provide a unified, simplified interface to a complex subsystem.
//!   The facade hides internal complexity and coordinates multiple services.
//!
//! When to use:
//!   - When a subsystem has many interacting types
//!   - When clients only need simple operations from a complex system
//!
//! Key Participants:
//!   - Facade: Simple API that delegates to subsystem types
//!   - Subsystem types: Do the actual work
//!   - Client: Uses only the facade

use std::collections::HashMap;

// ---------------------------------------------------------------------------
// Subsystem: Inventory
// ---------------------------------------------------------------------------
struct InventoryService {
    stock: HashMap<String, u32>,
}

impl InventoryService {
    fn new() -> Self {
        let mut stock = HashMap::new();
        stock.insert("LAPTOP-001".to_string(), 50);
        stock.insert("MOUSE-002".to_string(), 200);
        stock.insert("KEYBOARD-003".to_string(), 75);
        Self { stock }
    }

    fn reserve(&mut self, product_id: &str, quantity: u32) -> bool {
        let available = self.stock.get(product_id).copied().unwrap_or(0);
        if available >= quantity {
            self.stock.insert(product_id.to_string(), available - quantity);
            println!("  [Inventory] Reserved {}x {}", quantity, product_id);
            true
        } else {
            false
        }
    }
}

// ---------------------------------------------------------------------------
// Subsystem: Payment
// ---------------------------------------------------------------------------
struct PaymentService;

impl PaymentService {
    fn charge(&self, customer_id: &str, amount: f64) -> (bool, String) {
        println!(
            "  [Payment] Charged ${:.2} to customer {}",
            amount, customer_id
        );
        (true, format!("PAY-{}", rand_id()))
    }
}

// ---------------------------------------------------------------------------
// Subsystem: Shipping
// ---------------------------------------------------------------------------
struct ShippingService;

impl ShippingService {
    fn calculate_cost(&self, weight_kg: f64, _destination: &str) -> f64 {
        5.99 + weight_kg * 1.50
    }

    fn create_shipment(&self, order_id: &str, destination: &str) -> String {
        let tracking = format!("SHIP-{}", rand_id());
        println!(
            "  [Shipping] Created shipment for order {} to {}",
            order_id, destination
        );
        tracking
    }
}

// ---------------------------------------------------------------------------
// Subsystem: Notification
// ---------------------------------------------------------------------------
struct NotificationService;

impl NotificationService {
    fn send_email(&self, email: &str, subject: &str, _body: &str) {
        println!("  [Notification] Email to {}: {}", email, subject);
    }
}

fn rand_id() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_nanos() as u64
}

// ---------------------------------------------------------------------------
// Facade
// ---------------------------------------------------------------------------
struct OrderFacade {
    inventory: InventoryService,
    payment: PaymentService,
    shipping: ShippingService,
    notification: NotificationService,
}

#[derive(Debug)]
struct OrderResult {
    success: bool,
    order_id: Option<String>,
    total: Option<f64>,
    tracking: Option<String>,
    error: Option<String>,
}

impl OrderFacade {
    fn new() -> Self {
        Self {
            inventory: InventoryService::new(),
            payment: PaymentService,
            shipping: ShippingService,
            notification: NotificationService,
        }
    }

    fn place_order(
        &mut self,
        customer_id: &str,
        email: &str,
        product_id: &str,
        quantity: u32,
        unit_price: f64,
        destination: &str,
    ) -> OrderResult {
        println!("Processing order for {}...", customer_id);

        if !self.inventory.reserve(product_id, quantity) {
            return OrderResult {
                success: false,
                error: Some("Product out of stock".to_string()),
                order_id: None,
                total: None,
                tracking: None,
            };
        }

        let shipping_cost = self
            .shipping
            .calculate_cost(quantity as f64 * 0.5, destination);
        let total = unit_price * quantity as f64 + shipping_cost;

        let (paid, _txn_id) = self.payment.charge(customer_id, total);
        if !paid {
            return OrderResult {
                success: false,
                error: Some("Payment failed".to_string()),
                order_id: None,
                total: None,
                tracking: None,
            };
        }

        let order_id = format!("ORD-{}", rand_id());
        let tracking = self.shipping.create_shipment(&order_id, destination);

        self.notification.send_email(
            email,
            "Order Confirmed!",
            &format!("Order {} — tracking: {}", order_id, tracking),
        );

        OrderResult {
            success: true,
            order_id: Some(order_id),
            total: Some(total),
            tracking: Some(tracking),
            error: None,
        }
    }
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
fn main() {
    let mut facade = OrderFacade::new();

    let result = facade.place_order(
        "CUST-42",
        "alice@example.com",
        "LAPTOP-001",
        2,
        999.99,
        "New York, NY",
    );
    println!("\nOrder result: {:?}", result);

    println!("\n--- Out of stock scenario ---");
    let result2 = facade.place_order(
        "CUST-99",
        "bob@example.com",
        "NONEXISTENT",
        1,
        49.99,
        "LA, CA",
    );
    println!("Order result: {:?}", result2);
}
